File: TS/系统/03．技能系统/00．技能模板+函数/04．机制组件/10．复杂战斗通用机制/03．召唤物组状态管理.ts
/** @noSelfInFile */

import type { 机制清理篮子 } from "../06．机制清理/01．机制清理篮子";

const jass = require("jass.common") as any;

const GetHandleId = jass.GetHandleId as (h: any) => number;
const RemoveUnit = jass.RemoveUnit as (unit: any) => void;

const { registerDeathListener, unregisterDeathListener } = require("系统.00．核心系统.01．事件中心.07．单位死亡事件中心") as {
  registerDeathListener: (this: void, callback: (this: void, dyingUnit: any, killingUnit: any) => void) => void;
  unregisterDeathListener: (this: void, callback: (this: void, dyingUnit: any, killingUnit: any) => void) => void;
};
const { addDelayedCallback, removeDelayedCallback } = require("系统.00．核心系统.05．中心计时器") as {
  addDelayedCallback: (this: void, delayMs: number, callback: (this: void) => void) => number;
  removeDelayedCallback: (this: void, id: number) => void;
};

export interface 召唤物组状态参数 {
  清理?: 机制清理篮子;
  名称: string;
  全灭延迟秒?: number;
  全灭后保留死亡记录?: boolean;
  on数量变化?: (this: void, 存活数量: number, 总登记数量: number) => void;
  on单位死亡?: (this: void, 单位: any, 击杀者: any, 组: 召唤物组状态) => void;
  on全部死亡?: (this: void, 组: 召唤物组状态) => void;
}

export interface 召唤物组状态 {
  登记(单位: any): void;
  移除(单位: any, 是否删除单位?: boolean): void;
  取存活数量(): number;
  取总登记数量(): number;
  取单位列表(): any[];
  清空(是否删除单位?: boolean): void;
  销毁(): void;
}

const 召唤物组表: Record<number, 召唤物组状态实现 | undefined> = {};
let 已注册死亡监听 = false;

function 确保死亡监听(this: void): void {
  if (已注册死亡监听) return;
  已注册死亡监听 = true;
  registerDeathListener(on召唤物死亡);
}

function 尝试停止死亡监听(this: void): void {
  for (const key in 召唤物组表) {
    if (召唤物组表[key] != null) return;
  }
  if (!已注册死亡监听) return;
  unregisterDeathListener(on召唤物死亡);
  已注册死亡监听 = false;
}

function on召唤物死亡(this: void, dyingUnit: any, killingUnit: any): void {
  if (dyingUnit == null || dyingUnit === 0) return;
  const id = GetHandleId(dyingUnit);
  for (const key in 召唤物组表) {
    const 组 = 召唤物组表[key];
    if (组 != null && 组.包含ID(id)) 组.处理死亡(dyingUnit, killingUnit);
  }
}

class 召唤物组状态实现 implements 召唤物组状态 {
  readonly ID: number;
  private 参数: 召唤物组状态参数;
  private 单位列表: any[] = [];
  private 死亡表: Record<number, true | undefined> = {};
  private 全灭延迟ID = 0;
  private 已销毁 = false;

  constructor(ID: number, 参数: 召唤物组状态参数) {
    this.ID = ID;
    this.参数 = 参数;
    召唤物组表[ID] = this;
    确保死亡监听();
  }

  登记(单位: any): void {
    if (this.已销毁 || 单位 == null || 单位 === 0) return;
    this.单位列表.push(单位);
    this.死亡表[GetHandleId(单位)] = undefined;
    this.广播数量变化();
  }

  移除(单位: any, 是否删除单位: boolean = false): void {
    if (单位 == null || 单位 === 0) return;
    const id = GetHandleId(单位);
    for (let i = this.单位列表.length - 1; i >= 0; i--) {
      if (GetHandleId(this.单位列表[i]) === id) this.单位列表.splice(i, 1);
    }
    delete this.死亡表[id];
    if (是否删除单位) RemoveUnit(单位);
    this.广播数量变化();
  }

  取存活数量(): number {
    let count = 0;
    for (let i = 0; i < this.单位列表.length; i++) {
      const unit = this.单位列表[i];
      if (unit != null && unit !== 0 && this.死亡表[GetHandleId(unit)] == null) count++;
    }
    return count;
  }

  取总登记数量(): number {
    return this.单位列表.length;
  }

  取单位列表(): any[] {
    const result: any[] = [];
    for (let i = 0; i < this.单位列表.length; i++) result.push(this.单位列表[i]);
    return result;
  }

  清空(是否删除单位: boolean = false): void {
    if (this.全灭延迟ID !== 0) {
      removeDelayedCallback(this.全灭延迟ID);
      this.全灭延迟ID = 0;
    }
    if (是否删除单位) {
      for (let i = 0; i < this.单位列表.length; i++) {
        const unit = this.单位列表[i];
        if (unit != null && unit !== 0) RemoveUnit(unit);
      }
    }
    this.单位列表 = [];
    this.死亡表 = {};
    this.广播数量变化();
  }

  销毁(): void {
    if (this.已销毁) return;
    this.已销毁 = true;
    this.清空(false);
    delete 召唤物组表[this.ID];
    尝试停止死亡监听();
  }

  包含ID(id: number): boolean {
    for (let i = 0; i < this.单位列表.length; i++) {
      if (GetHandleId(this.单位列表[i]) === id) return true;
    }
    return false;
  }

  处理死亡(单位: any, 击杀者: any): void {
    const id = GetHandleId(单位);
    if (this.死亡表[id] === true) return;
    this.死亡表[id] = true;
    if (this.参数.on单位死亡 != null) this.参数.on单位死亡(单位, 击杀者, this);
    this.广播数量变化();
    if (this.取存活数量() <= 0) this.调度全灭();
  }

  private 调度全灭(): void {
    if (this.全灭延迟ID !== 0) return;
    const self = this;
    const delay = (this.参数.全灭延迟秒 ?? 0) * 1000;
    this.全灭延迟ID = addDelayedCallback(delay, function 召唤物组全灭(this: void): void {
      self.全灭延迟ID = 0;
      if (self.取存活数量() > 0) return;
      if (self.参数.on全部死亡 != null) self.参数.on全部死亡(self);
      if (self.参数.全灭后保留死亡记录 !== true) self.清空(false);
    });
  }

  private 广播数量变化(): void {
    if (this.参数.on数量变化 != null) this.参数.on数量变化(this.取存活数量(), this.单位列表.length);
  }
}

let 下一个召唤物组ID = 0;

export function 创建召唤物组状态(this: void, 参数: 召唤物组状态参数): 召唤物组状态 {
  const 实例 = new 召唤物组状态实现(++下一个召唤物组ID, 参数);
  if (参数.清理 != null) {
    参数.清理.登记清理(参数.名称, function 召唤物组状态清理(this: void): void {
      实例.销毁();
    });
  }
  return 实例;
}

File: src/系统/03．技能系统/00．技能模板+函数/04．机制组件/10．复杂战斗通用机制/03．召唤物组状态管理.lua
local ____lualib = require("lualib_bundle")
local __TS__Class = ____lualib.__TS__Class
local __TS__ArraySplice = ____lualib.__TS__ArraySplice
local __TS__Delete = ____lualib.__TS__Delete
local __TS__New = ____lualib.__TS__New
local ____exports = {}
local ____on_53EC_5524_7269_6B7B_4EA1, GetHandleId, _____53EC_5524_7269_7EC4_8868
function ____on_53EC_5524_7269_6B7B_4EA1(dyingUnit, killingUnit)
    if dyingUnit == nil or dyingUnit == 0 then
        return
    end
    local id = GetHandleId(dyingUnit)
    for key in pairs(_____53EC_5524_7269_7EC4_8868) do
        local _____7EC4 = _____53EC_5524_7269_7EC4_8868[key]
        if _____7EC4 ~= nil and _____7EC4["包含ID"](_____7EC4, id) then
            _____7EC4["处理死亡"](_____7EC4, dyingUnit, killingUnit)
        end
    end
end
local jass = require("jass.common")
GetHandleId = jass.GetHandleId
local RemoveUnit = jass.RemoveUnit
local ____require_result_0 = require("系统.00．核心系统.01．事件中心.07．单位死亡事件中心")
local registerDeathListener = ____require_result_0.registerDeathListener
local unregisterDeathListener = ____require_result_0.unregisterDeathListener
local ____require_result_1 = require("系统.00．核心系统.05．中心计时器")
local addDelayedCallback = ____require_result_1.addDelayedCallback
local removeDelayedCallback = ____require_result_1.removeDelayedCallback
_____53EC_5524_7269_7EC4_8868 = {}
local _____5DF2_6CE8_518C_6B7B_4EA1_76D1_542C = false
local function _____786E_4FDD_6B7B_4EA1_76D1_542C()
    if _____5DF2_6CE8_518C_6B7B_4EA1_76D1_542C then
        return
    end
    _____5DF2_6CE8_518C_6B7B_4EA1_76D1_542C = true
    registerDeathListener(____on_53EC_5524_7269_6B7B_4EA1)
end
local function _____5C1D_8BD5_505C_6B62_6B7B_4EA1_76D1_542C()
    for key in pairs(_____53EC_5524_7269_7EC4_8868) do
        if _____53EC_5524_7269_7EC4_8868[key] ~= nil then
            return
        end
    end
    if not _____5DF2_6CE8_518C_6B7B_4EA1_76D1_542C then
        return
    end
    unregisterDeathListener(____on_53EC_5524_7269_6B7B_4EA1)
    _____5DF2_6CE8_518C_6B7B_4EA1_76D1_542C = false
end
local _____53EC_5524_7269_7EC4_72B6_6001_5B9E_73B0 = __TS__Class()
_____53EC_5524_7269_7EC4_72B6_6001_5B9E_73B0.name = "召唤物组状态实现"
function _____53EC_5524_7269_7EC4_72B6_6001_5B9E_73B0.prototype.____constructor(self, ID, _____53C2_6570)
    self["单位列表"] = {}
    self["死亡表"] = {}
    self["全灭延迟ID"] = 0
    self["已销毁"] = false
    self.ID = ID
    self["参数"] = _____53C2_6570
    _____53EC_5524_7269_7EC4_8868[ID] = self
    _____786E_4FDD_6B7B_4EA1_76D1_542C()
end
_____53EC_5524_7269_7EC4_72B6_6001_5B9E_73B0.prototype["登记"] = function(self, _____5355_4F4D)
    if self["已销毁"] or _____5355_4F4D == nil or _____5355_4F4D == 0 then
        return
    end
    local ____self__5355_4F4D_5217_8868_2 = self["单位列表"]
    ____self__5355_4F4D_5217_8868_2[#____self__5355_4F4D_5217_8868_2 + 1] = _____5355_4F4D
    self["死亡表"][GetHandleId(_____5355_4F4D)] = nil
    self["广播数量变化"](self)
end
_____53EC_5524_7269_7EC4_72B6_6001_5B9E_73B0.prototype["移除"] = function(self, _____5355_4F4D, _____662F_5426_5220_9664_5355_4F4D)
    if _____662F_5426_5220_9664_5355_4F4D == nil then
        _____662F_5426_5220_9664_5355_4F4D = false
    end
    if _____5355_4F4D == nil or _____5355_4F4D == 0 then
        return
    end
    local id = GetHandleId(_____5355_4F4D)
    do
        local i = #self["单位列表"] - 1
        while i >= 0 do
            if GetHandleId(self["单位列表"][i + 1]) == id then
                __TS__ArraySplice(self["单位列表"], i, 1)
            end
            i = i - 1
        end
    end
    __TS__Delete(self["死亡表"], id)
    if _____662F_5426_5220_9664_5355_4F4D then
        RemoveUnit(_____5355_4F4D)
    end
    self["广播数量变化"](self)
end
_____53EC_5524_7269_7EC4_72B6_6001_5B9E_73B0.prototype["取存活数量"] = function(self)
    local count = 0
    do
        local i = 0
        while i < #self["单位列表"] do
            local unit = self["单位列表"][i + 1]
            if unit ~= nil and unit ~= 0 and self["死亡表"][GetHandleId(unit)] == nil then
                count = count + 1
            end
            i = i + 1
        end
    end
    return count
end
_____53EC_5524_7269_7EC4_72B6_6001_5B9E_73B0.prototype["取总登记数量"] = function(self)
    return #self["单位列表"]
end
_____53EC_5524_7269_7EC4_72B6_6001_5B9E_73B0.prototype["取单位列表"] = function(self)
    local result = {}
    do
        local i = 0
        while i < #self["单位列表"] do
            result[#result + 1] = self["单位列表"][i + 1]
            i = i + 1
        end
    end
    return result
end
_____53EC_5524_7269_7EC4_72B6_6001_5B9E_73B0.prototype["清空"] = function(self, _____662F_5426_5220_9664_5355_4F4D)
    if _____662F_5426_5220_9664_5355_4F4D == nil then
        _____662F_5426_5220_9664_5355_4F4D = false
    end
    if self["全灭延迟ID"] ~= 0 then
        removeDelayedCallback(self["全灭延迟ID"])
        self["全灭延迟ID"] = 0
    end
    if _____662F_5426_5220_9664_5355_4F4D then
        do
            local i = 0
            while i < #self["单位列表"] do
                local unit = self["单位列表"][i + 1]
                if unit ~= nil and unit ~= 0 then
                    RemoveUnit(unit)
                end
                i = i + 1
            end
        end
    end
    self["单位列表"] = {}
    self["死亡表"] = {}
    self["广播数量变化"](self)
end
_____53EC_5524_7269_7EC4_72B6_6001_5B9E_73B0.prototype["销毁"] = function(self)
    if self["已销毁"] then
        return
    end
    self["已销毁"] = true
    self["清空"](self, false)
    __TS__Delete(_____53EC_5524_7269_7EC4_8868, self.ID)
    _____5C1D_8BD5_505C_6B62_6B7B_4EA1_76D1_542C()
end
_____53EC_5524_7269_7EC4_72B6_6001_5B9E_73B0.prototype["包含ID"] = function(self, id)
    do
        local i = 0
        while i < #self["单位列表"] do
            if GetHandleId(self["单位列表"][i + 1]) == id then
                return true
            end
            i = i + 1
        end
    end
    return false
end
_____53EC_5524_7269_7EC4_72B6_6001_5B9E_73B0.prototype["处理死亡"] = function(self, _____5355_4F4D, _____51FB_6740_8005)
    local id = GetHandleId(_____5355_4F4D)
    if self["死亡表"][id] == true then
        return
    end
    self["死亡表"][id] = true
    if self["参数"]["on单位死亡"] ~= nil then
        self["参数"]["on单位死亡"](_____5355_4F4D, _____51FB_6740_8005, self)
    end
    self["广播数量变化"](self)
    if self["取存活数量"](self) <= 0 then
        self["调度全灭"](self)
    end
end
_____53EC_5524_7269_7EC4_72B6_6001_5B9E_73B0.prototype["调度全灭"] = function(self)
    if self["全灭延迟ID"] ~= 0 then
        return
    end
    local ____self = self
    local delay = (self["参数"]["全灭延迟秒"] or 0) * 1000
    self["全灭延迟ID"] = addDelayedCallback(
        delay,
        function()
            ____self["全灭延迟ID"] = 0
            if ____self["取存活数量"](____self) > 0 then
                return
            end
            if ____self["参数"]["on全部死亡"] ~= nil then
                ____self["参数"]["on全部死亡"](____self)
            end
            if ____self["参数"]["全灭后保留死亡记录"] ~= true then
                ____self["清空"](____self, false)
            end
        end
    )
end
_____53EC_5524_7269_7EC4_72B6_6001_5B9E_73B0.prototype["广播数量变化"] = function(self)
    if self["参数"]["on数量变化"] ~= nil then
        self["参数"]["on数量变化"](
            self["取存活数量"](self),
            #self["单位列表"]
        )
    end
end
local _____4E0B_4E00_4E2A_53EC_5524_7269_7EC4ID = 0
____exports["创建召唤物组状态"] = function(_____53C2_6570)
    local ____53EC_5524_7269_7EC4_72B6_6001_5B9E_73B0_3 = _____53EC_5524_7269_7EC4_72B6_6001_5B9E_73B0
    _____4E0B_4E00_4E2A_53EC_5524_7269_7EC4ID = _____4E0B_4E00_4E2A_53EC_5524_7269_7EC4ID + 1
    local _____5B9E_4F8B = __TS__New(____53EC_5524_7269_7EC4_72B6_6001_5B9E_73B0_3, _____4E0B_4E00_4E2A_53EC_5524_7269_7EC4ID, _____53C2_6570)
    if _____53C2_6570["清理"] ~= nil then
        local ____self_4 = _____53C2_6570["清理"]
        ____self_4["登记清理"](
            ____self_4,
            _____53C2_6570["名称"],
            function()
                _____5B9E_4F8B["销毁"](_____5B9E_4F8B)
            end
        )
    end
    return _____5B9E_4F8B
end
return ____exports
